oracle_fdw.c

```diff
@@ -2928,37 +2986,143 @@ foreign_join_ok(PlannerInfo *root, RelOptInfo *joinrel, JoinType jointype,
 	 *
-	 * For an inner join, clauses from both the relations are added to the
+	 * For an INNER join, clauses from both the relations are added to the
 	 * other remote clauses.
 	 *
+	 * For LEFT and RIGHT OUTER join, the clauses from the outer side are added
+	 * to remote_conds since those can be evaluated after the join is evaluated.
+	 * The clauses from inner side are added to the joinclauses, since they
+	 * need to evaluated while constructing the join.
+	 *
+	 * For a FULL OUTER JOIN, the other clauses from either relation can not
+	 * be added to the joinclauses or remote_conds, since each relation acts
+	 * as an outer relation for the other. Consider such full outer join as
+	 * unshippable because of the reasons mentioned above in this comment.
+	 *
 	 * The joining sides can not have local conditions, thus no need to test
 	 * shippability of the clauses being pulled up.
 	 */
-	fdwState->remote_conds = list_concat(fdwState->remote_conds,
-								  list_copy(fdwState_i->remote_conds));
-	fdwState->remote_conds = list_concat(fdwState->remote_conds,
-								  list_copy(fdwState_o->remote_conds));
+	switch (jointype)
+	{
+		case JOIN_INNER:
+			elog(DEBUG1, "INNER");
+			fdwState->remote_conds = list_concat(fdwState->remote_conds,
+										  list_copy(fdwState_i->remote_conds));
+			fdwState->remote_conds = list_concat(fdwState->remote_conds,
+										  list_copy(fdwState_o->remote_conds));
+			break;
 
-	/*
-	 * For an inner join, all restrictions can be treated alike. Treating the
-	 * pushed down conditions as join conditions allows a top level full outer
-	 * join to be deparsed without requiring subqueries.
-	 */
-	fdwState->joinclauses = fdwState->remote_conds;
-	fdwState->remote_conds = NIL;
+		case JOIN_LEFT:
+			elog(DEBUG1, "LEFT");
+			fdwState->joinclauses = list_concat(fdwState->joinclauses,
+										  list_copy(fdwState_i->remote_conds));
+			fdwState->remote_conds = list_concat(fdwState->remote_conds,
+										  list_copy(fdwState_o->remote_conds));
+			break;
+
+		case JOIN_RIGHT:
+			elog(DEBUG1, "RIGHT");
+			fdwState->joinclauses = list_concat(fdwState->joinclauses,
+										  list_copy(fdwState_o->remote_conds));
+			fdwState->remote_conds = list_concat(fdwState->remote_conds,
+										  list_copy(fdwState_i->remote_conds));
+			break;
+
+		case JOIN_FULL:
+			elog(DEBUG1, "FULL");
+			if (fdwState_i->remote_conds || fdwState_o->remote_conds)
+			{
+				elog(DEBUG1, "*e*");
+				return false;
+			}
+			break;
+
+		default:
+			/* Should not happen, we have just check this above */
+			elog(ERROR, "unsupported join type %d", jointype);
+	}
+
+	if (IS_OUTER_JOIN(jointype))
+	{
+		StringInfoData where; /* for outer join's WHERE clause */
+		char *keyword = "WHERE";
+
+		/*
+		 * If outer join query has no joinclauses, it means CROSS JOIN so it is
+		 * not pushed down.
+		 */
+		if (fdwState->joinclauses == NIL)
+		{
+			elog(DEBUG1, "*f*");
+			return false;
+		}
```

Comment:
I think this check is too late, because directly above this, we add `remote_conds` from one of the base relations to `joinclauses`.

Look at this query:

    test=> SELECT a.id FROM typetest1 a LEFT JOIN (SELECT * FROM typetest1 WHERE vc = 'short') b ON TRUE;
    DEBUG:  oracle_fdw: plan foreign table scan
    DEBUG:  oracle_fdw: set NLS_LANG=AMERICAN_AMERICA.AL32UTF8
    DEBUG:  oracle_fdw: begin serializable remote transaction
    DEBUG:  oracle_fdw: plan foreign table scan
    DEBUG:  oracle_fdw: set NLS_LANG=AMERICAN_AMERICA.AL32UTF8
    DEBUG:  LEFT
    DEBUG:  joinclauses tmp: (r4."VC" = 'short')
    DEBUG:  foreign_join_ok fdwState->where_clause: (null)
    DEBUG:  fdwState->where_clause: (null)
    DEBUG:  oracle_fdw: remote query is: SELECT /*f0f2514ad43ff1467c3a0d34b9f3bfd0*/ r1."ID" FROM ("TYPETEST1" r1 LEFT JOIN "TYPETEST1" r4 ON (r4."VC" = 'short'))
    DEBUG:  oracle_fdw: begin foreign join
    DEBUG:  oracle_fdw: execute query in foreign table scan 
    DEBUG:  oracle_fdw: end foreign table scan
    DEBUG:  oracle_fdw: commit remote transaction
     id 
    ----
      1
      3
      4
    (3 rows)

We end up pushing down what is effectively a cross join.

Maybe that check should be in the same place where it is for inner joins.

Reply:
Thanks for revising comments and your suggest.
You are right.
I fixed that problem with 8e5a02d.
That check was moved to near extract_actual_join_clauses.
Is it Okey?

Reply:
Hmm... I now realized we would be better to investigate handling a sub query in a join query.
Because below explain doesn't show the alias t2 for sub query on inner side.
Fow now, I'm not sure how to handle sub query in join query.

	EXPLAIN (COSTS off) SELECT t1.id, t2.id FROM typetest1 t1 INNER JOIN (SELECT * FROM typetest1 WHERE vc = 'short') t2 ON true;
	                                                            QUERY PLAN
	----------------------------------------------------------------------------------------------------------------------------------
	 Nested Loop
	   ->  Foreign Scan on typetest1
	         Oracle query: SELECT /*ca8a876017325c446646472507d12799*/ r4."ID", r4."VC" FROM "TYPETEST1" r4 WHERE (r4."VC" = 'short')
	   ->  Foreign Scan on typetest1 t1
	         Oracle query: SELECT /*709894645e7b46ee7c82f53e6eeb7cac*/ r1."ID" FROM "TYPETEST1" r1
	(5 rows)

Reply:
Commit 8e5a02df2fb3c3dc547d69cb8019b170335c8b0f looks good, thanks.

I don't see the problem with the aliases.
There is no alias on the first foreign scan, because no alias was specified in the subquery.
The alias `t2` does not appear anywhere because it is an alias for the whole subquery.
It would be inappropriate and confusing to use the alias `t2` for the first foreign scan.
Besides, that outsode the scope of oracle_fdw, these lines are generated in the core code.

Reply:
Oh, I see.
Thanks!